File: linked_list.py
# --coding: utf8 --
"""
创建一个链表的，添加100个元素，每个元素是有符号的整型
赋值是随机的
取值范围是[-98, 97)

然后删掉其中的负数
"""
import random
import copy

class Node(object):
    """
    docstring
    """
    def __init__(self, val, _next):
        self.val = val
        self._next = _next


def gen_random():
    return random.randint(-98, 96)

def gen_node():
    val = gen_random()
    node = Node(val, None)
    return node


def gen_list():
    node_root = gen_node()
    node = node_root
    for i in range(99):
        new_node = gen_node()
        node._next = new_node
        node = new_node
    return node_root

def print_node(node):
    string = "["
    while True:
        if node._next == None:
            break
        string += str(node.val) + ","
        node = node._next
    string = string.rstrip(",")
    string += "]"
    print(string)

node = gen_list()
print_node(node)


def remove_negative(node_root):
    while True:
        if node_root._next is None:
            break
        if node_root.val > 0:
            break
        node_root = node_root._next
    node = node_root
    next_node = node._next
    while True:
        if next_node is None:
            print '---break next---', node.val, next_node.val
            node._next = None
            break
        if next_node._next is None:
            print '---break next next---', node.val, next_node.val
            # node._next = None
            next_node._next = None
            break
        if next_node.val < 0:
            next_node = next_node._next
            continue
        node._next = next_node
        node = next_node
        next_node = next_node._next
    return node_root

print("-----before delete--------")
node = remove_negative(node)
print_node(node)

File: linked_list.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
)

type Node struct {
	Val int
	Next *Node
}


func genRandom() int{
	// -98 ~ 97
	return int(rand.Int63n(195)) - 98
}

func genList(nodeRoot *Node) {
	idx := 99
	for {
		if idx <= 0 {
			break
		}
		node := Node{Val: genRandom()}
		nodeRoot.Next = &node
		nodeRoot = &node
		idx--
	}
	return
}

func printList(node *Node) {
	str := "["
	for {
		str = str + fmt.Sprintf("%d,", node.Val)
		if node.Next == nil {
			break
		}
		node = node.Next
	}
	str = strings.TrimRight(str, ",")
	str = str + "]"
	fmt.Println(str)
}

func removeNagetive(node *Node) *Node {
	nodeRoot := node
	for {
		if nodeRoot.Val > 0 {
			break
		}
		nodeRoot = nodeRoot.Next
	}
	if nodeRoot.Next == nil {
		return nodeRoot
	}
	nodeLast := nodeRoot
	nodeNext := nodeRoot.Next
	for {
		if nodeNext.Next == nil {
			nodeLast.Next = nil
			break
		}
		if nodeNext.Val > 0 {
			nodeLast.Next = nodeNext
			nodeLast = nodeNext
		}
		nodeNext = nodeNext.Next
	}
	return nodeRoot
}

func main() {
	val := genRandom()
	node := Node{Val: val}
	fmt.Println("---start----")
	genList(&node)
	printList(&node)
	fmt.Println("---start remove----")
	newNode := removeNagetive(&node)
	printList(newNode)
	fmt.Println("---end----")
}
